scylla-nodetool: adjust effective ownership handling

pwrobelse opened the pull request:
When a keyspace uses tablets, then effective ownership can be obtained per table. If the user passes only a keyspace, then /storage_service/ownership/{keyspace} returns an error.

This change:
 - adds an additional positional parameter to 'status' command that allows a user to query status for table in a keyspace
 - makes usage of /storage_service/ownership/{keyspace} optional to avoid errors when user tries to obtain effective ownership of a keyspace that uses tablets
 - implements new frontend tests in 'test_status.py' that verify the new logic

Refs: scylladb#17405

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7444/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7444/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8552/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/520/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7444/)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider4.cloudius-systems.com

denesb commented on test/nodetool/test_status.py:
```diff
@@ -201,9 +208,14 @@ def _do_test_status(nodetool, keyspace, node_list, resolve=None):
+        keyspaces_using_tablets = [keyspace] if keyspace_uses_tablets else []
+        expected_requests += [
+                expected_request("GET", "/storage_service/keyspaces", params={"replication": "tablets"}, multiple=expected_request.ANY, response=keyspaces_using_tablets),
+                expected_request("GET", f"/storage_service/ownership/{keyspace}", multiple=expected_request.ANY, response=ownership_response)]
```
Why are you using `ANY` here? There will be just one of each request, no?

pwrobelse replied:
`/storage_service/keyspaces` should be accessed just once.

`/storage_service/ownership/{keyspace}` is accessed one or zero times depending on `keyspaces_using_tablets`. I will adjust the code.

pwrobelse replied:
Fixed

pwrobelse replied:
Sadly, if I set the exact count, then tests run with `-nodetool=cassandra` fail. 

Cassandra's nodetool does not call `/storage_service/keyspace`, but always call `/storage_service/ownership/{keyspace}`.

So, the expectation would need to be set depending on the nodetool type used.

denesb replied:
But the cassandra nodetool shouldn't even execute this branch at all -- it doesn't have tablets support.

denesb replied:
Ah, this is the `table is None` branch.

denesb replied:
Maybe worth adding a branch over which nodetool variant is used here.

pwrobelse replied:
> Maybe worth adding a branch over which nodetool variant is used here.

Done, thanks :+1:

denesb commented on test/nodetool/test_status.py:
```diff
@@ -159,7 +159,14 @@ def _get_ownership(nodes):
+def _do_test_status(nodetool, keyspace_desc, node_list, resolve=None, table=None):
+    if keyspace_desc:
+        keyspace = keyspace_desc["keyspace"]
+        keyspace_uses_tablets = keyspace_desc["uses_tablets"]
```
Use a `NamedTuple` instead.

denesb replied:
(you can also add the `table` field to it, and remove the separate param)

pwrobelse replied:
Done via:
`StatusQueryTarget = namedtuple('StatusQueryTarget', ['keyspace', 'table', 'uses_tablets'])`

denesb replied:
A mode modern version is:
```python
from typing import NamedTuple

class StatusQueryTarget(NameTuple):
    keyspace: str
    table: str
    uses_tablets: bool
```
Notice, how this format allows you to also use type-hints.

denesb commented on test/nodetool/test_status.py:
```diff
@@ -297,7 +313,85 @@ def test_status_keyspace_single_dc(nodetool):
+    _do_test_status(nodetool=nodetool, keyspace_desc=keyspace_desc, node_list=nodes)
+
+
+@pytest.mark.parametrize("uses_tablets", (False, True))
```
Instead of duplicating these tests, you can just parametrize the existing ones, like you do here.

pwrobelse replied:
Fixed

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -1776,11 +1777,44 @@ void snapshot_operation(scylla_rest_client& client, const bpo::variables_map& vm
+    const std::unordered_map<sstring, sstring> params = {{"replication", "tablets"}};
+    const auto tablet_keyspaces = rjson_array_to_sstring_vector(client.get("/storage_service/keyspaces", params));
+    const auto found_it = std::find(std::begin(tablet_keyspaces), std::end(tablet_keyspaces), keyspace);
+    return found_it != std::end(tablet_keyspaces);
```
This can be done much simpler:
```c++
auto res = client.get("/storage_service/keyspaces", params);
auto ks_array = res.GetArray();
return std::find_if(ks_array.begin(), ks_array.end(), [&] (const auto& json_ks) { return rjson::to_string_view(json_ks) == keyspace; }) != ks_array.end();
```

pwrobelse replied:
Fixed. I also removed `rjson_array_to_sstring_vector()` - it is not needed now.

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -3290,6 +3326,7 @@ Fore more information, see: https://opensource.docs.scylladb.com/stable/operatin
                 },
                 {
                     typed_option<sstring>("keyspace", "The keyspace name", 1),
+                    typed_option<sstring>("table", "The column family name", 1),
```
Let's use the new terminology and also mention tablets: "The table name (needed to display load information, if the keyspace uses tablets)"

pwrobelse replied:
Fixed

pwrobelse commented:
The changes introduced via new patch:
1. The merge conflicts had been resolved.
2. Introduced usage of `NamedTuple` instead of `dict` to pass the description to `_do_test_status()`.
3. Removed code duplication via `@pytest.mark.parametrize`.
4. Made expected count of REST API requests more robust in `_do_test_status()` via setting the exact count depending on the type of nodetool that is used.
5. Simplified `static bool keyspace_uses_tablets()`.
6. Adjusted command line parameters description.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7626/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7626/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/624/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8656/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7626/)
#### Build Details:
- Duration: 1 hr 59 min
- Builder: spider4.cloudius-systems.com
